Fixes an IonStructToMap object inspector NPE

cheqianh opened the pull request:
## Description:

This PR adds a null check for IonStructToMap object inspector to avoid NPE.

This PR is consisted of two commits, the first one ([91736de](https://github.com/amzn/ion-hive-serde/pull/91/commits/91736de6c0c65822c403af9c7c13e53643f2a329)) contains unit tests and the second one ([a226202](https://github.com/amzn/ion-hive-serde/pull/91/commits/a22620293e0ff562d2e1e1c5b9bbe198ff774db2)) includes the fix. 

## Issue:
Ion-hive-serde Struct to Map conversion throws NPE for null value. 

More specifically,  ion-path-extractor only matches the top-level container instead of Ion values within the nested container with default configuration. So the null values within the nested container are not filtered and are still treated as Ion null other than primitive null as we expect [here](https://github.com/amzn/ion-hive-serde/blob/master/serde/src/main/java/com/amazon/ionhiveserde/configuration/PathExtractionConfig.java#L72-L76) (This is kind of like the case insensitive issue we saw earlier) so that Ion case insensitive decorator will later convert it back to primitive null [here](https://github.com/amzn/ion-hive-serde/blob/hive2/serde/src/main/java/com/amazon/ionhiveserde/caseinsensitivedecorator/IonCaseInsensitiveDecorator.java#L29-L31) when [iterate](https://github.com/amzn/ion-hive-serde/blob/hive2/serde/src/main/java/com/amazon/ionhiveserde/caseinsensitivedecorator/IonContainerCaseInsensitiveDecorator.java#L217) the struct. Therefor [in the for loop](https://github.com/amzn/ion-hive-serde/blob/hive2/serde/src/main/java/com/amazon/ionhiveserde/objectinspectors/IonStructToMapObjectInspector.java#L82), null.getFieldName() throws NPE. 

## Solution in this PR
Since Hive doesn't know how to parse Ion null so we need to add primitive null to the struct. Currently, we filter Ion null in [deserializer](https://github.com/amzn/ion-hive-serde/blob/master/serde/src/main/java/com/amazon/ionhiveserde/IonHiveSerDe.java#L114) before asking object inspector to do any conversion. But object inspector should be able to detect Ion null as it passed the final Map to upper layer (E.g. presto/hive). So In this PR, I added a Null check in object inspector to make sure all values are parsed correctly. 
 
The solution follows three points below.
1. Don't change ion-path-extractor's current behavior. We may change ion-path-extractor to filter null values within a nested struct but in this PR we leave path-extractor as is.
2. To avoid any confusion, Ion container case insensitive decorator should behavior the same as regular ion container without any null check logic, specifically for iterator. When decorator's iterator encounter an Ion null, it should returns Ion null rather than null.
3. Ion object inspector should be able to check null type independently. It's ok to have other methods to filter null but object inspectors should make sure all values are validated before passing them to hive/presto.


I added null check for Ion Map to Struct, once it's addressed. I'll add null check for other struct object inspectors. 
 



By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

cheqianh commented on serde/src/test/kotlin/com/amazon/ionhiveserde/configuration/CaseInsensitiveDecoratorTest.kt:
```diff
@@ -130,14 +131,33 @@ class CaseInsensitiveDecoratorTest {
+        val iter = struct.iterator()
+        val v = iter.next()
+        // Returns Ion null like what Ion container does
+        assertEquals(v, ION.newNull())
```
CaseInsensitiveDecorator's iterator should behavior the same as regular Ion container iterator. It should returns Ion null instead of null.

cheqianh commented on serde/src/test/kotlin/com/amazon/ionhiveserde/objectinspectors/IonStructToMapObjectInspectorTest.kt:
```diff
@@ -59,26 +60,43 @@ class IonStructToMapObjectInspectorTest {
         val struct = makeStruct()
-
         val actual = subject.getMap(struct)
-        assertEquals(3, actual.size)
```
Remove this line to be a separate unit test, see [getMapSize](https://github.com/amzn/ion-hive-serde/pull/91/commits/91736de6c0c65822c403af9c7c13e53643f2a329#diff-e277b22a3db092c7502625ccb36b032f85990fd78606a2c7becfb6684500b09bR64) above.

cheqianh commented on serde/src/main/java/com/amazon/ionhiveserde/caseinsensitivedecorator/IonCaseInsensitiveDecorator.java:
```diff
@@ -22,11 +22,11 @@
      */
     public static IonValue wrapValue(final IonValue v) {
-        if (v == null || v.isNullValue()) {
+        if (v == null) {
```
Should behavior the same as the regular Ion container's iterator (Don't need to return null here). Object inspector will validate null values.

jobarr-amzn replied:
Thanks for tracking this down :)

jobarr-amzn commented on serde/src/test/kotlin/com/amazon/ionhiveserde/configuration/CaseInsensitiveDecoratorTest.kt:
```diff
@@ -130,14 +131,33 @@ class CaseInsensitiveDecoratorTest {
 
+    @Test
+    fun ionStructCaseInsensitiveDecoratorIteratorNext() {
+        val struct = struct_for(" { Foo: 1, Bar: '2'}")
```
```suggestion
        val struct = struct_for("{ Foo: 1, Bar: '2' }")
```

cheqianh replied:
Addressed.

jobarr-amzn commented on serde/src/test/kotlin/com/amazon/ionhiveserde/configuration/CaseInsensitiveDecoratorTest.kt:
```diff
@@ -224,4 +244,24 @@ class CaseInsensitiveDecoratorTest {
+    fun ionSequenceCaseInsensitiveDecoratorIteratorNext() {
+        val sequence = sequence_for("[1, '2']")
+        val iter = sequence.iterator()
+        val foo = iter.next()
+        assertEquals(foo, ION.newInt(1))
+        val bar = iter.next()
+        assertEquals(bar, ION.newSymbol("2"))
```
```suggestion
        assertEquals(iter.next(), ION.newInt(1))
        assertEquals(iter.next(), ION.newSymbol("2"))
```

`foo` and `bar` aren't really adding anything here.

cheqianh replied:
Addressed.

jobarr-amzn commented on serde/src/test/kotlin/com/amazon/ionhiveserde/objectinspectors/IonStructToMapObjectInspectorTest.kt:
```diff
@@ -107,4 +122,13 @@ class IonStructToMapObjectInspectorTest {
 
-}
\ No newline at end of file
+    private fun makeCaseInsensitiveStruct(): IonStruct {
+        val rawStruct = ION.newEmptyStruct()
+        val struct = IonStructCaseInsensitiveDecorator(rawStruct)
+
+        struct.add("a", ION.newInt(1))
+        struct.add("b", ION.newNull())
+
+        return struct
+    }
+}
```
I'd rather see CaseInsensitiveDecoratorTest's `struct_for` extracted to a commonly-accessible place. That pattern is easier to read at a glance.

Same applies to `makeStruct()`, if you could :)

cheqianh replied:
Extracted to Util.kt - [220c1e7](https://github.com/amzn/ion-hive-serde/pull/91/commits/220c1e77356f0d30488e0cd5a86201e997b7f336)

jobarr-amzn commented on serde/src/main/java/com/amazon/ionhiveserde/objectinspectors/IonStructToMapObjectInspector.java:
```diff
@@ -80,12 +80,20 @@ public Object getMapValueElement(final Object data, final Object key) {
         final Map<String, IonValue> map = new HashMap<>(size, DEFAULT_LOAD_FACTOR);
         for (IonValue v : struct) {
-            map.put(v.getFieldName(), v);
+            if (!IonUtil.isIonNull(v)) {
+                map.put(v.getFieldName(), v);
+            }
```
Is this right? In `{ foo: null }`, the Ion null has a field name. Is `null` not a valid value in a map?

cheqianh replied:
Discussed offline.

jobarr-amzn commented on serde/src/test/kotlin/com/amazon/ionhiveserde/Util.kt:
```diff
@@ -15,7 +15,53 @@
+    return IonSequenceCaseInsensitiveDecorator(sequence_for(s))
+}
+
+/**
+ * Create a sample struct used for testing. To create a desired Ion struct, use struct_for().
+ */
+internal fun makeStruct(): IonStruct {
+    return struct_for("{a: 1, b: 2, c: 3}")
+}
```
We ought to inline this method. Helper methods are to reduce work for the reader and increase clarity- burying the details of the constructed struct inside this helper obscures but without a payoff to make it worth it. The direct `struct_for` invocation will do just fine.

cheqianh replied:
Got it, changed to inline for better read experience.

jobarr-amzn commented on serde/src/test/kotlin/com/amazon/ionhiveserde/Util.kt:
```diff
@@ -15,7 +15,53 @@
+    return v
+}
+
+internal fun case_insensitive_decorator_sequence_for(s : String): IonSequence {
+    return IonSequenceCaseInsensitiveDecorator(sequence_for(s))
+}
```
This could be replaced with the following, then inlined:

```kotlin
IonCaseInsensitiveDecorator.valueFor(sequence_for(s));
```

If that's too verbose to read in our tests then you can do this:

```kotlin
internal fun case_insensitive(v: IonValue) {
  IonCaseInsensitiveDecorator.wrapValue(v)
}

// use like this:
case_insensitive(struct_for("{ foo: 1, bar: 2 }"));
```

This reads nearly the same, but uses a more general mechanism.

cheqianh replied:
Addressed with the second option. 

`case_insensitive` is more generic and works for both struct and sequence.

jobarr-amzn commented on serde/src/test/kotlin/com/amazon/ionhiveserde/Util.kt:
```diff
@@ -15,7 +15,53 @@
+    return v
+}
+
+internal fun case_insensitive_decorator_struct_for(s : String): IonStruct {
+    return IonStructCaseInsensitiveDecorator(struct_for(s))
+}
```
Replace with inlined `wrapValue` or `case_insensitive` method as described below.

jobarr-amzn commented on serde/src/test/kotlin/com/amazon/ionhiveserde/Util.kt:
```diff
@@ -15,7 +15,53 @@
+    return struct_for("{a: 1, b: 2, c: 3}")
+}
+
+/**
+ * Create a sample case insensitive struct used for testing. To create a desired case insensitive Ion struct,
+ * use case_insensitive_decorator_struct_for().
+ */
+internal fun makeCaseInsensitiveStruct(): IonStruct {
+    return case_insensitive_decorator_struct_for("{a: 1, b: null}")
+}
```
This method should be removed.

jobarr-amzn commented on serde/src/test/kotlin/com/amazon/ionhiveserde/objectinspectors/IonStructToMapObjectInspectorTest.kt:
```diff
@@ -59,26 +60,40 @@ class IonStructToMapObjectInspectorTest {
     }
 
+    @Test
+    fun getMapSize() {
+        assertEquals(3, subject.getMapSize(makeStruct()))
+    }
```
This test makes me ask "WHY is 3 the right value?". Inlining `makeStruct()` will make the answer obvious.

cheqianh commented:
Above commit worked for unit test.

cheqianh commented:
We are currently removing null value from the top-level container [here](https://github.com/amzn/ion-hive-serde/blob/hive3/serde/src/main/java/com/amazon/ionhiveserde/configuration/PathExtractionConfig.java#L72-L76) since Hive can't handle IonNull. However this code chunk is incorrect; we need to keep IonNull value as primitive null instead of removing it from the container. This is a good opportunity to remove unnecessary null check logic and rely on object inspectors to filter Ion null. 

So new improved since last review below:
[1845a62](https://github.com/amzn/ion-hive-serde/pull/91/commits/1845a6202f9ae7c2b87b863a63118330195854ac) changed unit tests style for easier debugging. 
[d2bc9dc](https://github.com/amzn/ion-hive-serde/pull/91/commits/d2bc9dc0c793f06bb60ef6c03ddd7dd32d9ecade), [1fa0a61](https://github.com/amzn/ion-hive-serde/pull/91/commits/1fa0a61d4736b24703eb966a8a57dd6631f53dd8) and [9836da9](https://github.com/amzn/ion-hive-serde/pull/91/commits/9836da9bacea957f869fc700bcf1bdd73849095b) added null type check for Ion container object inspectors and their unit tests.

cheqianh commented on serde/src/main/java/com/amazon/ionhiveserde/configuration/PathExtractionConfig.java:
```diff
@@ -69,11 +69,8 @@ class PathExtractionConfig {
-                } else {
-                    struct.put(columnName, ionValue);
-                }
+                // Hive can't handle IonNull, and we will filter all IonNull later in object inspectors
+                struct.put(columnName, ionValue);
```
These lines removes Ion Null from the top-level container. We now rely on object inspector to filter Ion Null as we still wants null value to be inside the final container.

cheqianh commented on serde/src/main/java/com/amazon/ionhiveserde/objectinspectors/IonStructToMapObjectInspector.java:
```diff
@@ -82,18 +82,15 @@ public Object getMapValueElement(final Object data, final Object key) {
                 map.put(v.getFieldName(), v);
+            } else {
+                // Hive can't handle Ion null
+                map.put(v.getFieldName(), null);
```
Here is where NPE occurs initially, `v` shouldn't be primitive null here now.

cheqianh commented:
### Description:

We are currently removing null value from the top-level container [here](https://github.com/amzn/ion-hive-serde/blob/hive3/serde/src/main/java/com/amazon/ionhiveserde/configuration/PathExtractionConfig.java#L72-L76) since Hive can't handle IonNull. However this code chunk doesn't work as expected - we need to keep IonNull values as primitive null instead of removing them from the container. This is a good opportunity to remove unnecessary null check logic and rely on object inspectors to filter Ion null. 

### Changes since the last review are:

An overview of changes after last review is [here](https://github.com/amzn/ion-hive-serde/pull/91/files/220c1e77356f0d30488e0cd5a86201e997b7f336..80f0f252940316efcd9924d8accc6f64b8cd980a).

[1845a62](https://github.com/amzn/ion-hive-serde/pull/91/commits/1845a6202f9ae7c2b87b863a63118330195854ac) improved unit tests style for easier debugging. 
[d2bc9dc](https://github.com/amzn/ion-hive-serde/pull/91/commits/d2bc9dc0c793f06bb60ef6c03ddd7dd32d9ecade), [1fa0a61](https://github.com/amzn/ion-hive-serde/pull/91/commits/1fa0a61d4736b24703eb966a8a57dd6631f53dd8) and [9836da9](https://github.com/amzn/ion-hive-serde/pull/91/commits/9836da9bacea957f869fc700bcf1bdd73849095b) added null type check for Ion container object inspectors and their unit tests.  
[80f0f25](https://github.com/amzn/ion-hive-serde/pull/91/commits/80f0f252940316efcd9924d8accc6f64b8cd980a) makes object inspector's get value methods also return primitive null for IonNull. I don't know where these methods will be used, but I kept them behavior as before to avoid any issue.

tgregg commented on serde/src/test/kotlin/com/amazon/ionhiveserde/objectinspectors/IonStructToMapObjectInspectorTest.kt:
```diff
@@ -40,11 +41,11 @@ class IonStructToMapObjectInspectorTest {
         assertEquals(ION.newInt(1), subject.getMapValueElement(struct, ION.newSymbol("a")))
         assertEquals(ION.newInt(2), subject.getMapValueElement(struct, ION.newSymbol("b")))
-        assertEquals(ION.newInt(3), subject.getMapValueElement(struct, ION.newSymbol("c")))
+        assertEquals(null, subject.getMapValueElement(struct, ION.newSymbol("c")))
```
Does this test fail before your fix and succeed after it?

cheqianh replied:
Yeah, I fall back to the commit before this PR and only modify this unit test. It shows
<img width="939" alt="Screen Shot 2022-09-07 at 5 52 53 PM" src="https://user-images.githubusercontent.com/67451029/189009905-26f1b40a-92ec-4e1e-a81d-f02cb52b65c9.png">

cheqianh replied:
Let me also provide the before/after test results for the initial issue.

tgregg reviewed (approved):
Looks good

jobarr-amzn commented on serde/src/main/java/com/amazon/ionhiveserde/objectinspectors/IonStructToMapObjectInspector.java:
```diff
@@ -80,7 +81,12 @@ public Object getMapValueElement(final Object data, final Object key) {
 
         final Map<String, IonValue> map = new HashMap<>(size, DEFAULT_LOAD_FACTOR);
         for (IonValue v : struct) {
-            map.put(v.getFieldName(), v);
+            if (!IonUtil.isIonNull(v)) {
+                map.put(v.getFieldName(), v);
+            } else {
+                // Hive can't handle Ion null
+                map.put(v.getFieldName(), null);
```
With a common `handleNull` function, this can become:
```java
        return StreamSupport.stream(struct.spliterator(), false)
                .collect(HashMap::new, (m,v) -> m.put(v.getFieldName(), handleNull(v)), HashMap::putAll);
```

Beauty is in they eye of the beholder, this may not be preferred. `Collectors.toMap` would look clean but can't be used because it doesn't allow null values. 

If nothing else we can collapse the conditional.
```java
            map.put(v.getFieldName(), handleNull(v));
```

cheqianh replied:
Thanks for the example, addressed. I tried to find a workaround to avoid that NPE but couldn't.

jobarr-amzn reviewed (commented):
I left some comments on factoring out a common `handleNull` kind of function, but didn't bother to note every instance. I think you get the idea.

cheqianh commented on serde/src/main/java/com/amazon/ionhiveserde/objectinspectors/IonSequenceToListObjectInspector.java:
```diff
@@ -53,8 +56,8 @@ public Object getListElement(final Object data, final int index) {
-
-        return sequence.get(index);
+        IonValue v = sequence.get(index);
+        return isIonNull(v) ? null : v;
```
Addressed.

cheqianh commented:
overview is available here
